tree/bst.go

```diff
@@ -160,3 +160,117 @@ func (n *node) getBSTNodeWithValue(val int) *node {
 
 	return n.right.getBSTNodeWithValue(val)
 }
+
+// changes By Prerna 08/01/2022
+
+// 1. Is validate BST
+func (t *BinarySearchTree) ValidBST() {
+	if t.isBST() {
+		fmt.Println("Valid BST")
+
+	} else {
+		fmt.Println("Invalid BST")
+	}
+}
+
+func (t *BinarySearchTree) isBST() bool {
+	if t == nil {
+		return true
+	}
+	return t.root.chkNode(100, 100)
+}
+func (n *node) chkNode(max int, min int) bool {
+	if n == nil {
+		return true
+	}
+	if n.val > max && n.val < min {
+		return false
+	}
+	if n.left.chkNode(n.val, min) && n.right.chkNode(max, n.val) {
+		return true
+	}
+	return false
+
+}
+
+// 2. determine if the given BST is balance
+
+func (t *BinarySearchTree) IsBal() {
+	if t == nil {
+		fmt.Println("Balanced")
+		return
+	}
+	h := 0
+	h1 := t.root.left.GetHght(h)
+	h2 := t.root.right.GetHght(h)
+	if h1-h2 > 1 || h2-h1 > 1 {
+		fmt.Println("Un-balanced")
+		return
+	}
+	fmt.Println("Balanced")
+}
+
+func (n *node) GetHght(h int) int {
+	if n == nil {
+		return h
+	}
+	n.left.GetHght(h + 1)
```

Comment:
Save the return of this call in a variable and use it directly in the if comparison. Also for the next line.

`lh := n.left.GetHght(h + 1)`
 `rh := n.right.GetHght(h + 1)`
`if lh>rh {
 return lh+1
}`

`return rh+1`

Something like this.